File: packages/filters-webgl/src/gradientGlowFilter.ts
import type { WebGLRenderTarget } from '@flighthq/render-webgl';
import type { GradientGlowFilter } from '@flighthq/types';
import type { WebGLRenderState } from '@flighthq/types';

import { applyBoxBlurFilterToWebGL } from './blurFilter';
import type { WebGLFilterLocations } from './filterPass';
import { clearWebGLFilterTarget, compileWebGLFilterProgram, drawWebGLFilterPass } from './filterPass';
import { createWebGLGradientRampTexture } from './gradientRamp';
import { applyWebGLBlitPass, applyWebGLTintPass } from './tintShader';

// Uses the blurred alpha (unit 0) to index into a gradient ramp texture (unit 1).
// Outputs the gradient-colored glow at the correct intensity per pixel.
const GRADIENT_LOOKUP_FRAGMENT_SRC = `#version 300 es
precision mediump float;
in vec2 v_texCoord;
uniform sampler2D u_texture;
uniform sampler2D u_ramp;
out vec4 fragColor;
void main() {
  float alpha = texture(u_texture, v_texCoord).a;
  fragColor = texture(u_ramp, vec2(alpha, 0.5));
}`;

type GradientLookupLocations = WebGLFilterLocations & {
  locRamp: WebGLUniformLocation;
};

const lookupShaders = new WeakMap<WebGLRenderState, GradientLookupLocations>();

/**
 * Applies a gradient glow to `source`, writing the result to `dest`.
 * The gradient ramp is built each call from `filter.colors`, `filter.alphas`,
 * and `filter.ratios` — cache the filter object to amortize this cost, or
 * pre-build a texture with `createWebGLGradientRampTexture` and draw with the
 * lower-level shader directly.
 *
 * Compositing order: gradient glow → source on top.
 *
 * `scratch` must contain three render targets of the same dimensions as `dest`.
 * The filter allocates a temporary `WebGLTexture` internally on each call.
 */
export function applyGradientGlowFilterToWebGL(
  state: WebGLRenderState,
  source: WebGLRenderTarget,
  dest: WebGLRenderTarget,
  scratch: WebGLRenderTarget[],
  filter: Readonly<Omit<GradientGlowFilter, 'type'>>,
): void {
  const quality = Math.max(1, Math.round(filter.quality ?? 1));
  const strength = filter.strength ?? 1;

  const [s0, s1, s2] = scratch;
  const gl = state.gl;

  // Extract alpha as a neutral (white) mask, then blur → s1
  applyWebGLTintPass(state, source, s0, 0xffffff, 1, Math.min(1, strength));
  applyBoxBlurFilterToWebGL(state, s0, s1, s2, { blurX: filter.blurX ?? 6, blurY: filter.blurY ?? 6, passes: quality });

  // Build gradient ramp texture and look up the blurred alpha → s0
  const ramp = createWebGLGradientRampTexture(gl, filter.colors, filter.alphas, filter.ratios);
  applyGradientLookupPass(state, s1, ramp, s0);
  gl.deleteTexture(ramp);

  clearWebGLFilterTarget(state, dest);
  applyWebGLBlitPass(state, s0, dest);
  applyWebGLBlitPass(state, source, dest);
}

function applyGradientLookupPass(
  state: WebGLRenderState,
  blurred: WebGLRenderTarget,
  ramp: WebGLTexture,
  dest: WebGLRenderTarget,
): void {
  const loc = getLookupShader(state);
  drawWebGLFilterPass(state, blurred, dest, loc, (gl) => {
    gl.activeTexture(gl.TEXTURE1);
    gl.bindTexture(gl.TEXTURE_2D, ramp);
    gl.uniform1i(loc.locRamp, 1);
    gl.activeTexture(gl.TEXTURE0);
  });
}

function getLookupShader(state: WebGLRenderState): GradientLookupLocations {
  let loc = lookupShaders.get(state);
  if (loc === undefined) {
    const gl = state.gl;
    const base = compileWebGLFilterProgram(gl, GRADIENT_LOOKUP_FRAGMENT_SRC);
    loc = { ...base, locRamp: gl.getUniformLocation(base.program, 'u_ramp')! };
    lookupShaders.set(state, loc);
  }
  return loc;
}

File: crates/flighthq-filters-wgpu/src/gradient_glow_filter.rs
//! Gradient-glow wgpu filter pass.
//!
//! Compositing order: gradient glow → source on top.
//!
//! `scratch` must contain three render targets of the same dimensions as `dest`.
//! This function allocates a temporary gradient ramp texture internally on each
//! call; it is destroyed before the function returns.

use flighthq_filters::GradientGlowFilter;
use flighthq_render_wgpu::{WgpuRenderState, WgpuRenderTarget};

use crate::blur_filter::apply_box_blur_filter_to_wgpu;
use crate::filter_pass::{
    WgpuBlendMode, WgpuFilterState, clear_wgpu_filter_target, create_wgpu_dual_source_pipeline,
    draw_wgpu_views_pass,
};
use crate::gradient_ramp::create_wgpu_gradient_ramp_texture;
use crate::tint_shader::{apply_wgpu_blit_pass, apply_wgpu_tint_pass};

// Uses the blurred alpha (group 1) to index into a gradient ramp texture (group 2).
const GRADIENT_LOOKUP_WGSL: &str = r#"
struct Uniforms { _u : f32, _pad0 : f32, _pad1 : f32, _pad2 : f32, }
@group(0) @binding(0) var<uniform> uni : Uniforms;
@group(1) @binding(0) var texBlurred : texture_2d<f32>;
@group(1) @binding(1) var smp : sampler;
@group(2) @binding(0) var texRamp : texture_2d<f32>;
@group(2) @binding(1) var smp2 : sampler;

@fragment
fn fs_main(@location(0) uv : vec2f) -> @location(0) vec4f {
  let alpha = textureSampleLevel(texBlurred, smp, uv, 0.0).a;
  return textureSampleLevel(texRamp, smp2, vec2f(alpha, 0.5), 0.0);
}"#;

/// Applies a gradient glow to `source`, writing the result to `dest`.
///
/// The gradient ramp is built each call from `filter.colors`, `filter.alphas`,
/// and `filter.ratios`. Compositing order: gradient glow → source on top.
///
/// `scratch` must contain three render targets of the same dimensions as `dest`.
pub fn apply_gradient_glow_filter_to_wgpu(
    state: &mut WgpuRenderState,
    filter_state: &mut WgpuFilterState,
    source: &WgpuRenderTarget,
    dest: &WgpuRenderTarget,
    scratch: &[&WgpuRenderTarget; 3],
    filter: &GradientGlowFilter,
) {
    let quality = filter.quality.unwrap_or(1).max(1);
    let strength = filter.strength.unwrap_or(1.0);

    let [s0, s1, s2] = *scratch;

    apply_wgpu_tint_pass(
        state,
        filter_state,
        source,
        s0,
        0xffffff,
        1.0,
        strength.min(1.0),
    );
    apply_box_blur_filter_to_wgpu(
        state,
        filter_state,
        s0,
        s1,
        s2,
        filter.blur_x.unwrap_or(6.0),
        filter.blur_y.unwrap_or(6.0),
        quality,
    );

    let ratios: Vec<u8> = filter
        .ratios
        .iter()
        .map(|&r| r.round().clamp(0.0, 255.0) as u8)
        .collect();
    let ramp = create_wgpu_gradient_ramp_texture(state, &filter.colors, &filter.alphas, &ratios);
    let ramp_view = ramp.create_view(&wgpu::TextureViewDescriptor::default());

    if filter_state.gradient_lookup_pipeline.is_none() {
        let p = create_wgpu_dual_source_pipeline(
            state,
            filter_state,
            GRADIENT_LOOKUP_WGSL,
            WgpuBlendMode::Premul,
        );
        filter_state.gradient_lookup_pipeline = Some(p);
    }
    let mut pipeline = filter_state.gradient_lookup_pipeline.take().unwrap();
    // blurred alpha (group 1) + gradient ramp (group 2) -> gradient glow into s0.
    draw_wgpu_views_pass(
        state,
        filter_state,
        &[&s1.view, &ramp_view],
        Some(s0),
        &mut pipeline,
        |_| {},
    );
    filter_state.gradient_lookup_pipeline = Some(pipeline);

    // Destroy before submit (deferred by wgpu until the encoder's work completes).
    ramp.destroy();

    clear_wgpu_filter_target(state, dest);
    apply_wgpu_blit_pass(state, filter_state, s0, dest);
    apply_wgpu_blit_pass(state, filter_state, source, dest);
}

#[cfg(test)]
mod tests {
    use super::*;

    // GRADIENT_LOOKUP_WGSL

    #[test]
    fn gradient_lookup_wgsl_indexes_ramp_by_blurred_alpha() {
        assert!(
            GRADIENT_LOOKUP_WGSL
                .contains("let alpha = textureSampleLevel(texBlurred, smp, uv, 0.0).a")
        );
        assert!(
            GRADIENT_LOOKUP_WGSL
                .contains("textureSampleLevel(texRamp, smp2, vec2f(alpha, 0.5), 0.0)")
        );
    }
}
